conda/gateways/repodata/__init__.py

```diff
@@ -289,3 +304,291 @@ def conda_http_errors(url, repodata_fn):
+        )
+
+    @property
+    def cache_path_state(self):
+        """
+        Out-of-band etag and other state needed by the RepoInterface.
+        """
+        return pathlib.Path(
+            self.cache_dir,
+            self.name + ("1" if context.use_only_tar_bz2 else "") + ".state.json",
+        )
+
+    def load(self) -> str:
+        # read state and repodata.json with locking
+
+        # lock .state.json
+        # read .state.json
+        # read repodata.json
+        # check stat, if wrong clear cache information
+
+        with self.cache_path_state.open("r+") as state_file, _lock(state_file):
+            # cannot use pathlib.read_text / write_text on any locked file, as
+            # it will release the lock early
+            state = json.loads(state_file.read())
+
+            # json and state files should match
+
+            # must read json before checking stat
+            json_data = self.cache_path_json.read_text()
+            json_stat = self.cache_path_json.stat()
+            if not (
+                state.get("mtime_ns") == json_stat.st_mtime_ns
+                and state.get("size") == json_stat.st_size
+            ):
+                # clear mod, etag, cache_control to encourage re-download
+                state.update({"etag": "", "mod": "", "cache_control": "", "size": 0})
+            self.state.clear()
+            self.state.update(
+                state
+            )  # will aliased _mod, _etag (not cleared above) pass through as mod, etag?
+
+        return json_data
+
+        # check repodata.json stat(); mtime_ns must equal .state.json, or it is stale
+        # read repodata.json
+        # check repodata.json stat() again: st_size, st_mtime_ns must be equal
+
+        # repodata.json is okay - use it somewhere
+
+        # repodata.json is not okay - maybe use it, but don't allow cache updates
+
+        # unlock .state.json
+
+        # also, add mtime_refreshed_ns instead of touching file
+
+    def save(self, data):
+        """
+        Write data to <repodata>.json cache path, synchronize state.
+        """
+        # contextmanager
+
+        # lock .state.json
+        # read
+        # write to temp path
+        # get mtime, etc. from temp path in cache directory
+        # move temp path on top of normal cache json
+        # write .state.json
+        # (in context manager: delete temp path)
+
+        temp_path = self.cache_dir / self.cache_path_base.with_suffix(
+            "." + os.urandom(4).hex() + ".tmp"
+        )
+
+        with temp_path.open("wx") as temp:  # exclusive mode, error if exists
+            temp.write(data)
+
+    def replace(self, temp_path: Path):
+        """
+        Rename path onto <repodata>.json path, synchronize state.
+
+        Relies on path's mtime not changing on move.
+        """
+        with self.cache_path_state.open("w+") as state_file, _lock(state_file):
```

Comment:
does `w+` (open for writing and reading) truncate the file, before bothering to check if we can get the lock?